chia/full_node/full_node_api.py

```diff
@@ -1720,5 +1712,201 @@ def get_fee_estimates(est: FeeEstimatorInterface, req_times: List[uint64]) -> Li
+        # It can be increased in the future, and this protocol should be written and tested in a way that
+        # this increase would not break the API.
+        count = 15000
+        puzzle_hashes = request.puzzle_hashes[:count]
+
+        (coin_states, next_min_height) = await self.full_node.coin_store.batch_coin_states_by_puzzle_hashes(
+            puzzle_hashes,
+            min_height=min_height,
+            include_spent=request.filters.include_spent,
+            include_unspent=request.filters.include_unspent,
+            include_hinted=request.filters.include_hinted,
+            min_amount=request.filters.min_amount,
+            max_items=max_items,
+        )
+        is_done = next_min_height is None
+
+        peak_height = self.full_node.blockchain.get_peak_height()
+        assert peak_height is not None
+
+        height = uint32(next_min_height - 1) if next_min_height is not None else peak_height
+        header_hash = self.full_node.blockchain.height_to_hash(height)
+        assert header_hash is not None
+
+        if is_done and request.subscribe_when_finished:
+            added = list(
+                self.full_node.subscriptions.add_puzzle_subscriptions(
+                    peer.peer_node_id, puzzle_hashes, max_subscriptions
+                )
+            )
+        else:
+            added = puzzle_hashes
+
+        response = wallet_protocol.RespondPuzzleState(added, height, header_hash, is_done, coin_states)
+        msg = make_msg(ProtocolMessageTypes.respond_puzzle_state, response)
+        return msg
+
+    @api_request(peer_required=True, reply_types=[ProtocolMessageTypes.respond_coin_state])
+    async def request_coin_state(self, request: wallet_protocol.RequestCoinState, peer: WSChiaConnection) -> Message:
+        max_items = self.max_subscribe_response_items(peer)
+        max_subscriptions = self.max_subscriptions(peer)
+
+        previous_header_hash = (
+            self.full_node.blockchain.height_to_hash(request.previous_height)
+            if request.previous_height is not None
+            else self.full_node.blockchain.constants.GENESIS_CHALLENGE
+        )
+        assert previous_header_hash is not None
+
+        if request.header_hash != previous_header_hash:
+            rejection = wallet_protocol.RejectCoinState()
+            msg = make_msg(ProtocolMessageTypes.reject_coin_state, rejection)
+            return msg
+
+        min_height = uint32(request.previous_height + 1 if request.previous_height is not None else 0)
+
+        coin_states = await self.full_node.coin_store.get_coin_states_by_ids(
+            True,
+            set(request.coin_ids),
+            min_height=min_height,
+            max_items=max_items,
+        )
+
+        if request.subscribe:
+            peer_id = peer.peer_node_id
+            added = list(
+                self.full_node.subscriptions.add_coin_subscriptions(peer_id, request.coin_ids, max_subscriptions)
```

Comment:
it would seem reasonable to only subscribe to the coin IDs that were included in the response.

Reply:
This would be an argument for `RequestAddCoinSubscriptions` then, it would be nice to have a way to subscribe to something which doesn't exist yet.

Reply:
I changed my mind on this. I agree that it would be good to be able to subscribe to coins that don't exist yet